File: grida-apis/apis-security/src/main/java/org/grida/util/UserIdResolver.java
package org.grida.util;

import org.grida.exception.ApisSecurityException;
import org.springframework.core.MethodParameter;
import org.springframework.web.bind.support.WebDataBinderFactory;
import org.springframework.web.context.request.NativeWebRequest;
import org.springframework.web.method.support.HandlerMethodArgumentResolver;
import org.springframework.web.method.support.ModelAndViewContainer;

import javax.servlet.http.HttpServletRequest;

import static org.grida.exception.ApisSecurityErrorCode.NO_AUTHENTICATED_USER;

public class UserIdResolver implements HandlerMethodArgumentResolver {

    private static final String USER_ID_ATTRIBUTE_KEY = "userId";

    @Override
    public boolean supportsParameter(MethodParameter parameter) {
        boolean hasAnnotation = parameter.hasParameterAnnotation(RequestUserId.class);
        boolean isLongValue = Long.class.isAssignableFrom(parameter.getParameterType());

        return hasAnnotation && isLongValue;
    }

    @Override
    public Object resolveArgument(MethodParameter parameter,
                                  ModelAndViewContainer mavContainer,
                                  NativeWebRequest webRequest,
                                  WebDataBinderFactory binderFactory) throws Exception {

        HttpServletRequest request = (HttpServletRequest) webRequest.getNativeRequest();
        Object userId = request.getAttribute(USER_ID_ATTRIBUTE_KEY);
        validateHasUserId(userId);
        return userId;
    }

    private void validateHasUserId(Object userId) {
        if (userId == null) {
            throw new ApisSecurityException(NO_AUTHENTICATED_USER);
        }
    }
}

File: grida-core/core-api/src/main/kotlin/org/grida/support/UserIdResolver.kt
package org.grida.support

import org.springframework.core.MethodParameter
import org.springframework.security.core.context.SecurityContextHolder
import org.springframework.web.bind.support.WebDataBinderFactory
import org.springframework.web.context.request.NativeWebRequest
import org.springframework.web.method.support.HandlerMethodArgumentResolver
import org.springframework.web.method.support.ModelAndViewContainer

class UserIdResolver : HandlerMethodArgumentResolver {

    override fun supportsParameter(parameter: MethodParameter): Boolean {
        return parameter.hasParameterAnnotation(RequestUserId::class.java)
    }

    override fun resolveArgument(
        parameter: MethodParameter,
        mavContainer: ModelAndViewContainer?,
        webRequest: NativeWebRequest,
        binderFactory: WebDataBinderFactory?
    ): Any {
        val securityContextHolder = SecurityContextHolder.getContext()
        val authentication = securityContextHolder.authentication

        return (authentication.principal as String).toLong()
    }
}
